libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,388 @@
+/*
+ * AVL Tree implementation.
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan 
+{
+template <typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable 
+{
+protected:
+    struct Node {
+        T* data = UAVCAN_NULLPTR;
+        int16_t h = 1; // initially added as leaf
+        Node* left = UAVCAN_NULLPTR;
+        Node* right = UAVCAN_NULLPTR;
+        Node* equal_keys = UAVCAN_NULLPTR;
+    };
+
+    Node* root_;
+
+private:
+    size_t len_ ;
+
+    static int16_t heightOf(const Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return n->h;
+    }
+
+    void postOrderTraverseRecursively(Node* n, std::function<void(T*&)> forEach) {
+        if (n == UAVCAN_NULLPTR) {
+            return;
+        }
+
+        postOrderTraverseRecursively(n->left, forEach);
+        postOrderTraverseRecursively(n->right, forEach);
+        forEach(n->data);
+    }
+
+    Node* makeNode(T* payload) {
+        void* praw = this->allocator_.allocate(sizeof(Node));
+
+        if (praw == UAVCAN_NULLPTR) {
+            UAVCAN_TRACE("AvlTree", " OOM -- Can't allocate Node");
+            return UAVCAN_NULLPTR; // Push rejected
+        }
+
+        Node* node = new (praw) Node();
+        UAVCAN_ASSERT(node);
+
+        node->data = payload;
+        return node;
+    }
+
+    void deleteNode(Node*& n) {
+        if (n != UAVCAN_NULLPTR) {
+            n->~Node();
+            allocator_.deallocate(n);
+            n = UAVCAN_NULLPTR;
+        }
+    }
+
+    static int16_t balanceOf(Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        const int32_t diff = heightOf(n->left) - heightOf(n->right);
+        return static_cast<int16_t>(std::max(-2, std::min(2, diff)));
```

Comment:
Given that this fails for `lpc11c24`, `note:   mismatched types 'std::initializer_list<_Tp>' and 'int'
         return static_cast<int16_t>(std::max(-2, std::min(2, diff)));`

Do we keep the last version?

Reply:
it should work when you  switch back to your own versions of minOf/maxOf. Its failing because of changes to stl (see: @pavel-kirienko  is almost always right)

Reply:
Yes, but do I need to keep the `maxOf` & `minOf` as-is, or move it to the `templates.hpp`?

Reply:
You can keep them here.